From 5bea066c7fe68b664d96be9c9c31301891ee0ff5 Mon Sep 17 00:00:00 2001
From: Tiejun Chen <tiejun.chen@windriver.com>
Date: Wed, 11 Jul 2012 13:26:13 +0800
Subject: [PATCH 479/547] powerpc/hugetlb: replace __get_cpu_var with
 get_cpu_var

Replace __get_cpu_var safely with get_cpu_var to avoid
the following call trace:

BUG: using smp_processor_id() in preemptible [00000000] code: hugetlbperforma/1288
caller is free_hugepd_range.constprop.22+0x90/0x1a4
Call Trace:
[e905dd60] [c0007b34] show_stack+0x10c/0x1c0 (unreliable)
[e905ddb0] [c0648d10] dump_stack+0x24/0x34
[e905ddc0] [c033a9b4] debug_smp_processor_id+0xd8/0xfc
[e905dde0] [c0015ef0] free_hugepd_range.constprop.22+0x90/0x1a4
[e905de00] [c001631c] hugetlb_free_pgd_range+0x7c/0x17c
[e905de40] [c00fa7ac] free_pgtables+0x110/0x12c
[e905de60] [c01003f4] unmap_region+0x108/0x134
[e905dee0] [c0101568] do_munmap+0x264/0x2e4
[e905df10] [c010162c] vm_munmap+0x44/0x68
[e905df30] [c0102c14] sys_munmap+0x1c/0x2c
[e905df40] [c000e33c] ret_from_syscall+0x0/0x3c

Signed-off-by: Tiejun Chen <tiejun.chen@windriver.com>
Signed-off-by: Bin Jiang <bin.jiang@windriver.com>
---
 arch/powerpc/mm/hugetlbpage.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/arch/powerpc/mm/hugetlbpage.c b/arch/powerpc/mm/hugetlbpage.c
index 77fdd2c..06b0470 100644
--- a/arch/powerpc/mm/hugetlbpage.c
+++ b/arch/powerpc/mm/hugetlbpage.c
@@ -518,12 +518,13 @@ static void hugepd_free(struct mmu_gather *tlb, void *hugepte)
 {
 	struct hugepd_freelist **batchp;
 
-	batchp = &__get_cpu_var(hugepd_freelist_cur);
+	batchp = &get_cpu_var(hugepd_freelist_cur);
 
 	if (atomic_read(&tlb->mm->mm_users) < 2 ||
 	    cpumask_equal(mm_cpumask(tlb->mm),
 			  cpumask_of(smp_processor_id()))) {
 		kmem_cache_free(hugepte_cache, hugepte);
+	put_cpu_var(hugepd_freelist_cur);
 		return;
 	}
 
@@ -537,6 +538,7 @@ static void hugepd_free(struct mmu_gather *tlb, void *hugepte)
 		call_rcu_sched(&(*batchp)->rcu, hugepd_free_rcu_callback);
 		*batchp = NULL;
 	}
+    put_cpu_var(hugepd_freelist_cur);
 }
 #endif
 
-- 
1.8.4.93.g57e4c17

